File: ruby/linked_list/2_add_two_numbers.rb
# Definition for singly-linked list.
class ListNode
  attr_accessor :val, :next
  def initialize(val = 0, _next = nil)
    @val = val
    @next = _next
  end

  def to_a
    result = []
    current = self
    until current.nil?
      result.append(current.val)
      current = current.next
    end
    result
  end
end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
  l2 if l1.nil?
  l1 if l2.nil?

  head = nil
  prev = nil
  carrying = 0
  while !l1.nil? || !l2.nil? || carrying.positive?
    sum = (l1&.val || 0) + (l2&.val || 0) + carrying
    carrying = 0
    carrying, sum = sum.divmod(10) if sum >= 10

    node = ListNode.new(sum)
    head = node if head.nil?

    prev.next = node unless prev.nil?
    prev = node

    l1 = l1&.next
    l2 = l2&.next
  end

  head
end

def populate_list(values = [])
  return nil if values.empty?

  head = ListNode.new(values[0])
  current = head

  values[1..].each do |value|
    nxt = ListNode.new(value)
    current.next = nxt

    current = nxt
  end

  head
end

def test
  # Test cases

  # result: [7, 0, 8]
  l1 = populate_list([2, 4, 3])
  l2 = populate_list([5, 6, 4])
  puts add_two_numbers(l1, l2).to_a.inspect

  # result: [0]
  l1 = populate_list([0])
  l2 = populate_list([0])
  puts add_two_numbers(l1, l2).to_a.inspect

  # result: [8, 9, 9, 9, 0, 0, 0, 1]
  l1 = populate_list([9, 9, 9, 9, 9, 9, 9])
  l2 = populate_list([9, 9, 9, 9])
  puts add_two_numbers(l1, l2).to_a.inspect
end

test

File: go/linked_list/2_add_two_numbers.go
package main

import "fmt"

func main() {
	// Test cases

	// result: [7, 0, 8]
	l1 := populateList([]int{2, 4, 3})
	l2 := populateList([]int{5, 6, 4})
	printList(addTwoNumbers(l1, l2))

	// result: [0]
	l1 = populateList([]int{0})
	l2 = populateList([]int{0})
	printList(addTwoNumbers(l1, l2))

	// result: [8, 9, 9, 9, 0, 0, 0, 1]
	l1 = populateList([]int{9, 9, 9, 9, 9, 9, 9})
	l2 = populateList([]int{9, 9, 9, 9})
	printList(addTwoNumbers(l1, l2))
}

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	if l1 == nil {
		return l2
	}
	if l2 == nil {
		return l1
	}

	var head *ListNode
	var prev *ListNode

	carrying := 0
	for l1 != nil || l2 != nil || carrying > 0 {
		sum := carrying
		if l1 != nil {
			sum += l1.Val
		}
		if l2 != nil {
			sum += l2.Val
		}
		carrying = 0
		if sum >= 10 {
			sum -= 10
			carrying = 1
		}

		node := ListNode{sum, nil}
		if head == nil {
			head = &node
		}
		if prev != nil {
			prev.Next = &node
		}
		prev = &node

		if l1 != nil {
			l1 = l1.Next
		} else {
			l1 = nil
		}
		if l2 != nil {
			l2 = l2.Next
		} else {
			l2 = nil
		}
	}
	return head
}

func populateList(values []int) *ListNode {
	if len(values) == 0 {
		return nil
	}

	head := ListNode{values[0], nil}
	current := &head

	for _, value := range values[1:] {
		next := ListNode{value, nil}
		current.Next = &next
		current = current.Next
	}

	return &head
}

func printList(head *ListNode) {
	current := head
	for current != nil {
		fmt.Println(current.Val)
		current = current.Next
	}
}
